src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -27,70 +28,364 @@
+  // not to be found. This makes searches for free space more efficient.  For each partition p, _leftmosts[p]
+  // represents its least index, and its _rightmosts[p] its greatest index. Empty intervals are indicated by the
+  // canonical [_max, -1].
+  ssize_t _leftmosts[UIntNumPartitions];
+  ssize_t _rightmosts[UIntNumPartitions];
+
+  // Allocation for humongous objects needs to find regions that are entirely empty.  For each partion p, _leftmosts_empty[p]
+  // represents the first region belonging to this partition that is completely empty and _rightmosts_empty[p] represents the
+  // last region that is completely empty.  If there is no completely empty region in this partition, this is represented
+  // by the canonical [_max, -1].
+  ssize_t _leftmosts_empty[UIntNumPartitions];
+  ssize_t _rightmosts_empty[UIntNumPartitions];
+
+  // For each partition p, _capacity[p] represents the total amount of memory within the partition at the time
+  // of the most recent rebuild, _used[p] represents the total amount of memory that has been allocated within this
+  // partition (either already allocated as of the rebuild, or allocated since the rebuild).  _capacity[p] and _used[p]
+  // are denoted in bytes.  Note that some regions that had been assigned to a particular partition at rebuild time
+  // may have been retired following the rebuild.  The tallies for these regions are still reflected in _capacity[p]
+  // and _used[p], even though the region may have been removed from the free set.
+  size_t _capacity[UIntNumPartitions];
+  size_t _used[UIntNumPartitions];
+  size_t _region_counts[UIntNumPartitions];
+
+  // Shrink the intervals associated with partition when region idx is removed from this free set
+  inline void shrink_interval_if_boundary_modified(ShenandoahFreeSetPartitionId partition, ssize_t idx);
+
+  // Shrink the intervals associated with partition when regions low_idx through high_idx inclusive are removed from this free set
+  inline void shrink_interval_if_range_modifies_either_boundary(ShenandoahFreeSetPartitionId partition,
+                                                                ssize_t low_idx, ssize_t high_idx);
+  inline void expand_interval_if_boundary_modified(ShenandoahFreeSetPartitionId partition, ssize_t idx, size_t capacity);
+
+  void dump_bitmap_row(ssize_t idx) const;
+  void dump_bitmap_range(ssize_t start_idx, ssize_t end_idx) const;
+  void dump_bitmap_all() const;
+
+public:
+  ShenandoahRegionPartitions(size_t max_regions, ShenandoahFreeSet* free_set);
+  ~ShenandoahRegionPartitions() {}
+
+  // Remove all regions from all partitions and reset all bounds
+  void make_all_regions_unavailable();
+
+  // Set the partition id for a particular region without adjusting interval bounds or usage/capacity tallies
+  inline void raw_set_membership(size_t idx, ShenandoahFreeSetPartitionId p) {
```

Comment:
I realize that the word `set` can be a little ambiguous here (from the mathematical notion of a set as  a collection of things), although the fact that it has a void return type tells us that `set` is the imperative verb here, not a collection of things. May be `raw_assign_membership(size_t region_idx, ShFSPId p)` if you want to, but I may be splitting hairs here :-)

Reply:
Thanks.  Making this change.